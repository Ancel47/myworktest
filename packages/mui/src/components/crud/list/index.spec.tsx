import React, { ReactNode } from "react";
import { Route, Routes } from "react-router-dom";
import { AccessControlProvider } from "@pankod/refine-core";
import { Table, TableRow } from "@mui/material";

import { act, render, TestWrapper } from "@test";
import { List } from "./index";

const renderList = (
    list: ReactNode,
    accessControlProvider?: AccessControlProvider,
) => {
    return render(
        <Routes>
            <Route path="/:resource" element={list} />
        </Routes>,
        {
            wrapper: TestWrapper({
                routerInitialEntries: ["/posts"],
                accessControlProvider,
            }),
        },
    );
};

describe("<List/>", () => {
    beforeEach(() => {
        // This is an issue on `mui` side rather than `refine`. Ignoring for now but might need to be fixed.
        jest.spyOn(console, "error").mockImplementation((message) => {
            if (message?.includes?.("validateDOMNesting")) {
                return;
            }

            console.warn(message);
        });
    });

    describe("JSON Rest Server", () => {
        it("mounts with table", async () => {
            jest.useFakeTimers();

            const { getByText } = renderList(
                <List key="posts">
                    <Table key={"id"}>No Data</Table>
                </List>,
            );

            await act(async () => {
                jest.advanceTimersToNextTimer(1);
            });

            getByText("No Data");
        });
        it("renders given data", async () => {
            jest.useFakeTimers();

            const { container } = renderList(
                <List key="posts">
                    <Table key="id">
                        <TableRow key="title" />
                    </Table>
                </List>,
            );

            await act(async () => {
                jest.advanceTimersToNextTimer(1);
            });

            expect(container).toMatchSnapshot();
        });

        it("should render optional title with title prop", async () => {
            jest.useFakeTimers();

            const { getByText } = renderList(
                <List
                    cardHeaderProps={{
                        title: "New Title",
                    }}
                ></List>,
            );

            await act(async () => {
                jest.advanceTimersToNextTimer(1);
            });

            getByText("New Title");
        });

        it("should render with label instead of resource name successfully", async () => {
            jest.useFakeTimers();

            const { getByText } = render(
                <Routes>
                    <Route path="/:resource" element={<List />} />
                </Routes>,
                {
                    wrapper: TestWrapper({
                        resources: [
                            {
                                name: "posts",
                                options: { route: "posts", label: "test" },
                            },
                        ],
                        routerInitialEntries: ["/posts"],
                    }),
                },
            );

            await act(async () => {
                jest.advanceTimersToNextTimer(1);
            });

            getByText("Tests");
        });

        describe("render create button", () => {
            it("should create edit button", async () => {
                jest.useFakeTimers();

                const { getByText, queryByTestId } = render(
                    <Routes>
                        <Route path="/:resource" element={<List />} />
                    </Routes>,
                    {
                        wrapper: TestWrapper({
                            resources: [
                                {
                                    name: "posts",
                                    create: () => null,
                                },
                            ],
                            routerInitialEntries: ["/posts"],
                        }),
                    },
                );

                await act(async () => {
                    jest.advanceTimersToNextTimer(1);
                });

                expect(queryByTestId("list-create-button")).not.toBeNull();

                getByText("Posts");
            });

            it("should not render create button on resource canCreate false", async () => {
                jest.useFakeTimers();

                const { getByText, queryByTestId } = render(
                    <Routes>
                        <Route path="/:resource" element={<List />} />
                    </Routes>,
                    {
                        wrapper: TestWrapper({
                            resources: [
                                {
                                    name: "posts",
                                },
                            ],
                            routerInitialEntries: ["/posts"],
                        }),
                    },
                );

                await act(async () => {
                    jest.advanceTimersToNextTimer(1);
                });

                expect(queryByTestId("list-create-button")).toBeNull();

                getByText("Posts");
            });

            it("should render create button on resource canCreate false & createButtonProps props not null on component", async () => {
                jest.useFakeTimers();

                const { getByText, queryByTestId } = render(
                    <Routes>
                        <Route
                            path="/:resource"
                            element={
                                <List createButtonProps={{ size: "large" }} />
                            }
                        ></Route>
                    </Routes>,
                    {
                        wrapper: TestWrapper({
                            routerInitialEntries: ["/posts"],
                        }),
                    },
                );

                await act(async () => {
                    jest.advanceTimersToNextTimer(1);
                });

                expect(queryByTestId("list-create-button")).not.toBeNull();

                getByText("Posts");
            });

            it("should render create button on resource canCreate true & createButtonProps props not null on component", async () => {
                jest.useFakeTimers();

                const { getByText, queryByTestId } = render(
                    <Routes>
                        <Route
                            path="/:resource"
                            element={
                                <List createButtonProps={{ size: "large" }} />
                            }
                        ></Route>
                    </Routes>,
                    {
                        wrapper: TestWrapper({
                            resources: [
                                {
                                    name: "posts",
                                    create: () => null,
                                },
                            ],
                            routerInitialEntries: ["/posts"],
                        }),
                    },
                );

                await act(async () => {
                    jest.advanceTimersToNextTimer(1);
                });

                expect(queryByTestId("list-create-button")).not.toBeNull();

                getByText("Posts");
            });

            it("should not render create button on resource canCreate true & canCreate props false on component", async () => {
                jest.useFakeTimers();

                const { queryByTestId } = render(
                    <Routes>
                        <Route
                            path="/:resource"
                            element={<List canCreate={false} />}
                        ></Route>
                    </Routes>,
                    {
                        wrapper: TestWrapper({
                            resources: [
                                {
                                    name: "posts",
                                    create: () => null,
                                },
                            ],
                            routerInitialEntries: ["/posts"],
                        }),
                    },
                );

                await act(async () => {
                    jest.advanceTimersToNextTimer(1);
                });

                expect(queryByTestId("list-create-button")).toBeNull();
            });

            it("should render create button on resource canCreate false & canCreate props true on component", async () => {
                jest.useFakeTimers();

                const { queryByTestId } = render(
                    <Routes>
                        <Route
                            path="/:resource"
                            element={<List canCreate={true} />}
                        ></Route>
                    </Routes>,
                    {
                        wrapper: TestWrapper({
                            resources: [
                                {
                                    name: "posts",
                                },
                            ],
                            routerInitialEntries: ["/posts"],
                        }),
                    },
                );

                await act(async () => {
                    jest.advanceTimersToNextTimer(1);
                });

                expect(queryByTestId("list-create-button")).not.toBeNull();
            });

            it("should render disabled create button if user doesn't have permission", async () => {
                jest.useFakeTimers();

                const { queryByTestId } = renderList(
                    <List canCreate={true} />,
                    {
                        can: ({ action }) => {
                            switch (action) {
                                case "create":
                                    return Promise.resolve({ can: false });
                                default:
                                    return Promise.resolve({ can: false });
                            }
                        },
                    },
                );

                await act(async () => {
                    jest.advanceTimersToNextTimer(1);
                });

                expect(queryByTestId("list-create-button")).toBeDisabled();
            });
            describe("Breadcrumb", () => {
                it("should not render breadcrumb", async () => {
                    jest.useFakeTimers();
                    const { queryByLabelText } = render(
                        <Routes>
                            <Route path="/:resource" element={<List />}></Route>
                        </Routes>,
                        {
                            wrapper: TestWrapper({
                                resources: [
                                    {
                                        name: "posts",
                                    },
                                ],
                                routerInitialEntries: ["/posts"],
                            }),
                        },
                    );

                    await act(async () => {
                        jest.advanceTimersToNextTimer(1);
                    });

                    expect(
                        queryByLabelText("breadcrumb"),
                    ).not.toBeInTheDocument();
                });
            });
        });
    });
});
