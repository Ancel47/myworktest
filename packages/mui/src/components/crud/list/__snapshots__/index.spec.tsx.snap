// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List/> JSON Rest Server renders given data 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
        >
          Posts
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        />
      </table>
    </div>
  </div>
</div>
`;
